homeassistant/components/webdav/backup.py

```diff
@@ -0,0 +1,269 @@
+"""Support for WebDAV backup."""
+
+from __future__ import annotations
+
+from collections.abc import AsyncIterator, Callable, Coroutine
+from functools import wraps
+import logging
+from typing import Any, Concatenate
+
+from aiohttp import ClientTimeout
+from aiowebdav2 import Property, PropertyRequest
+from aiowebdav2.exceptions import UnauthorizedError, WebDavError
+from propcache.api import cached_property
+
+from homeassistant.components.backup import (
+    AgentBackup,
+    BackupAgent,
+    BackupAgentError,
+    BackupNotFound,
+    suggested_filename,
+)
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.json import json_dumps
+from homeassistant.util.json import json_loads_object
+
+from . import WebDavConfigEntry
+from .const import CONF_BACKUP_PATH, DATA_BACKUP_AGENT_LISTENERS, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+METADATA_VERSION = "1"
+BACKUP_TIMEOUT = ClientTimeout(connect=10, total=43200)
+
+
+async def async_get_backup_agents(
+    hass: HomeAssistant,
+) -> list[BackupAgent]:
+    """Return a list of backup agents."""
+    entries: list[WebDavConfigEntry] = hass.config_entries.async_loaded_entries(DOMAIN)
+    return [WebDavBackupAgent(hass, entry) for entry in entries]
+
+
+@callback
+def async_register_backup_agents_listener(
+    hass: HomeAssistant,
+    *,
+    listener: Callable[[], None],
+    **kwargs: Any,
+) -> Callable[[], None]:
+    """Register a listener to be called when agents are added or removed.
+
+    :return: A function to unregister the listener.
+    """
+    hass.data.setdefault(DATA_BACKUP_AGENT_LISTENERS, []).append(listener)
+
+    @callback
+    def remove_listener() -> None:
+        """Remove the listener."""
+        hass.data[DATA_BACKUP_AGENT_LISTENERS].remove(listener)
+        if not hass.data[DATA_BACKUP_AGENT_LISTENERS]:
+            del hass.data[DATA_BACKUP_AGENT_LISTENERS]
+
+    return remove_listener
+
+
+def handle_backup_errors[_R, **P](
+    func: Callable[Concatenate[WebDavBackupAgent, P], Coroutine[Any, Any, _R]],
+) -> Callable[Concatenate[WebDavBackupAgent, P], Coroutine[Any, Any, _R]]:
+    """Handle backup errors."""
+
+    @wraps(func)
+    async def wrapper(self: WebDavBackupAgent, *args: P.args, **kwargs: P.kwargs) -> _R:
+        try:
+            return await func(self, *args, **kwargs)
+        except UnauthorizedError as err:
+            raise BackupAgentError("Authentication error") from err
+        except WebDavError as err:
+            _LOGGER.error(
```

Comment:
Include the message in the exception instead. It will be logged by the backup manager.